File: java/src/LeastRecentlyUsedCache.java
import java.util.HashMap;
import java.util.Map;

/**
 * Implementation of a Least Recently Used (LRU) cache using a HashMap for O(1) lookups
 * and a doubly-linked list for maintaining the order of items by their usage.
 */
public class LeastRecentlyUsedCache {
	private final int capacity;
	private final Map<String, LeastRecentlyUsedCacheItem> cache;
	private LeastRecentlyUsedCacheItem head;  // Most recently used
	private LeastRecentlyUsedCacheItem tail;  // Least recently used

	/**
	 * Creates a new LRU cache with the specified capacity.
	 * @param capacity The maximum number of items the cache can hold
	 * @throws IllegalArgumentException if capacity is less than 1
	 */
	public LeastRecentlyUsedCache(int capacity) {
		if (capacity < 1) {
			throw new IllegalArgumentException("Cache capacity must be at least 1");
		}
		this.capacity = capacity;
		this.cache = new HashMap<>();
	}

	/**
	 * Adds or updates a value in the cache. If the key already exists, its value is updated
	 * and it becomes the most recently used item. If the cache is full, the least recently
	 * used item is removed before adding the new item.
	 * @param key The key to store the value under
	 * @param value The value to store
	 */
	public void put(String key, Object value) {
		// If key exists, update it and move to front
		if (cache.containsKey(key)) {
			LeastRecentlyUsedCacheItem oldItem = cache.get(key);
			removeFromList(oldItem);  // Remove the old item from the list
			LeastRecentlyUsedCacheItem newItem = new LeastRecentlyUsedCacheItem(key, value);
			addToFront(newItem);
			cache.put(key, newItem);
			return;
		}

		// If cache is full, remove least recently used item
		if (cache.size() >= capacity) {
			removeLeastRecentlyUsed();
		}

		// Add new item
		LeastRecentlyUsedCacheItem newItem = new LeastRecentlyUsedCacheItem(key, value);
		addToFront(newItem);
		cache.put(key, newItem);
	}

	/**
	 * Retrieves a value from the cache. If the key exists, the item becomes the most
	 * recently used item. If the key doesn't exist, throws an exception.
	 * @param key The key to look up
	 * @return The value associated with the key
	 * @throws IllegalArgumentException if the key is not found in the cache
	 */
	public Object get(String key) {
		LeastRecentlyUsedCacheItem item = cache.get(key);
		if (item == null) {
			throw new IllegalArgumentException("Key not found in cache: " + key);
		}

		// Move to front of list (most recently used)
		removeFromList(item);
		addToFront(item);
		return item.getValue();
	}

	// Private helper methods

	private void removeLeastRecentlyUsed() {
		if (tail != null) {
			cache.remove(tail.getKey());
			removeFromList(tail);
		}
	}

	private void removeFromList(LeastRecentlyUsedCacheItem item) {
		if (item == null) return;

		// Update prev item's next pointer
		if (item.getPrev() != null) {
			item.getPrev().setNext(item.getNext());
		} else {
			// This was the head
			head = item.getNext();
		}

		// Update next item's prev pointer
		if (item.getNext() != null) {
			item.getNext().setPrev(item.getPrev());
		} else {
			// This was the tail
			tail = item.getPrev();
		}

		// Clear the item's pointers
		item.setNext(null);
		item.setPrev(null);
	}

	private void addToFront(LeastRecentlyUsedCacheItem item) {
		if (head == null) {
			// First item in the list
			head = item;
			tail = item;
		} else {
			// Add to front of list
			item.setNext(head);
			head.setPrev(item);
			head = item;
		}
	}
} 

File: ruby/src/least_recently_used_cache.rb
# frozen_string_literal: true

require_relative 'least_recently_used_cache_item'

# Implementation of a Least Recently Used (LRU) cache using a Hash for O(1) lookups
# and a doubly-linked list for maintaining the order of items by their usage.
class LeastRecentlyUsedCache
  def initialize(capacity)
    raise ArgumentError, 'Cache capacity must be at least 1' if capacity < 1

    @capacity = capacity
    @cache = {}
    @head = nil  # Most recently used
    @tail = nil  # Least recently used
  end

  # Adds or updates a value in the cache. If the key already exists, its value is updated
  # and it becomes the most recently used item. If the cache is full, the least recently
  # used item is removed before adding the new item.
  def put(key, value)
    if @cache.key?(key)
      # If key exists, update it and move to front
      old_item = @cache[key]
      remove_from_list(old_item)
      new_item = LeastRecentlyUsedCacheItem.new(key, value)
      add_to_front(new_item)
      @cache[key] = new_item
      return
    end

    # If cache is full, remove least recently used item
    remove_least_recently_used if @cache.size >= @capacity

    # Add new item
    new_item = LeastRecentlyUsedCacheItem.new(key, value)
    add_to_front(new_item)
    @cache[key] = new_item
  end

  # Retrieves a value from the cache. If the key exists, the item becomes the most
  # recently used item. If the key doesn't exist, raises an ArgumentError.
  def get(key)
    item = @cache[key]
    raise ArgumentError, "Key not found in cache: #{key}" if item.nil?

    # Move to front of list (most recently used)
    remove_from_list(item)
    add_to_front(item)
    item.value
  end

  private

  def remove_least_recently_used
    return if @tail.nil?

    @cache.delete(@tail.key)
    remove_from_list(@tail)
  end

  def remove_from_list(item)
    return if item.nil?

    # Update prev item's next pointer
    if item.prev
      item.prev.next = item.next
    else
      # This was the head
      @head = item.next
    end

    # Update next item's prev pointer
    if item.next
      item.next.prev = item.prev
    else
      # This was the tail
      @tail = item.prev
    end

    # Clear the item's pointers
    item.next = nil
    item.prev = nil
  end

  def add_to_front(item)
    if @head.nil?
      # First item in the list
      @head = item
      @tail = item
    else
      # Add to front of list
      item.next = @head
      @head.prev = item
      @head = item
    end
  end
end
